mbuild/lib/recipes/polymer.py

```diff
@@ -182,25 +194,60 @@ def build(self, n, sequence="A", add_hydrogens=True):
 
         head_tail = [head_port, tail_port]
 
-        for i, compound in enumerate(self._end_groups):
-            if compound is not None:
-                if self._headtail[i] is not None:
-                    head_tail[i].update_separation(self._headtail[i])
-                self.add(compound)
-                force_overlap(compound, compound.labels["up"], head_tail[i])
-            else:
-                if add_hydrogens:
-                    hydrogen = H()
-                    # Defaut to 1/2 H-C bond len
-                    head_tail[i].update_separation(0.0547)
-                    hydrogen["up"].update_separation(0.0547)
-                    self.add(hydrogen)
-                    force_overlap(hydrogen, hydrogen["up"], head_tail[i])
-                else:
-                    # if None, hoist port to polymer level
-                    self.add(
-                        head_tail[i], self._port_labels[i], containment=False
+        if make_periodic:
+            if np.any([x is not None for x in self._end_groups]):
+                raise ValueError(
+                    "Keyword 'make_periodic' cannot be defined if 'end_groups' are provided."
+                )
+            z_axis_transform(
+                compound=self,
+                point_on_z_axis=head_port.pos,
+                point_on_zx_plane=tail_port.pos,
+            )
+            bond_length = 2 * np.max(
+                [
+                    np.sqrt(np.sum(np.square(atm1.pos - atm2.pos)))
+                    for (atm1, atm2) in self.root.bond_graph.edges(
+                        self["monomer[0]"]
                     )
+                ]
+            )
+            self.periodic = [
```

Comment:
We could use periodicity, as it is written here, that may have been my intention when I first wrote it. I could also make it an internal variable for the error check so that it's not exposed.

Until the polymer is built, I'm not sure how the periodicity could be set, so this would take care of that for you. I will admit that I haven't used or set the periodicity when using mbuild since I usually make a large molecule, add a box around it, and solvate it with packmol.

Reply:
Ah that's fair. Periodicity is typically used when doing the simulation side of things anyways, so most of the time it's set in the control files (which is external to MoSDeF). In mBuild, it's mostly used when tiling out compounds, you can tile them out based on their periodicity. However, it is an attribute of a compound that recipes and other building methods can take advantage of. So it's possible to have some use case that would build a polymer, and then do something with that periodicity, which would be overwritten here. That's why I'm a little hesitant to have it modified here, so I agree an internal variable may be the way to go.